pkgs/applications/version-management/p4/default.nix

```
@@ -1,31 +1,108 @@
+        stripRoot = false;
+      };
+    in linkFarm "android-zlib-contrib" [
+      # We only want to keep the contrib directory as the other files conflict
+      # with p4's own zlib files. (For the same reason, we can't use the
+      # cone-based Git sparse checkout, either.)
+      { name = "contrib"; path = "${src}/contrib"; }
+    ];
+  libcxxUnified = symlinkJoin {
+    inherit (libcxx) name;
+    paths = [ libcxx libcxxabi ];
+  };
+
+  version = "2022.1.2305383";
+in
+stdenv.mkDerivation {
   pname = "p4";
-  version = "2021.2.2201121";
+  inherit version;
 
   src = fetchurl {
-    # actually  https://cdist2.perforce.com/perforce/r21.2/bin.linux26x86_64/helix-core-server.tgz but upstream deletes releases
-    url = "https://web.archive.org/web/20211118024943/https://cdist2.perforce.com/perforce/r21.2/bin.linux26x86_64/helix-core-server.tgz";
-    sha256 = "sha256-SrfI2ZD7KDyttCd8+fo8g4UZKljYYO/SbzqrS9tAcC8=";
+    # Upstream replaces minor versions, so use cached URL.
+    url = "https://web.archive.org/web/20220901184735/https://ftp.perforce.com/perforce/r22.1/bin.tools/p4source.tgz";
+    sha256 = "27ab3ddd7b178b05cf0b710e941650dac0688d294110ebafda9027732c0944c6";
   };
 
-  sourceRoot = ".";
+  buildInputs = lib.optionals stdenv.isDarwin [ CoreServices Foundation Security ];
+  nativeBuildInputs = [ jam ];
+
+  hardeningDisable = lib.optionals stdenv.isDarwin [ "strictoverflow" ];
 
-  dontBuild = true;
+  # Flags understood by the Jamfile and imported from the environment:
+  CROSS_COMPILE = stdenv.cc.targetPrefix;
+  CCFLAGS =
+    # The file contrib/optimizations/slide_hash_neon.h is missing from the
+    # upstream distribution. It comes from the Android/Chromium sources.
+    lib.optionals stdenv.isAarch64 [ "-I${androidZlibContrib}" ];
+  "C++FLAGS" =
+    # Avoid a compilation error that only occurs for 4-byte longs.
+    lib.optionals stdenv.isi686 [ "-Wno-narrowing" ]
+    # See the "Header dependency changes" section of
+    # https://www.gnu.org/software/gcc/gcc-11/porting_to.html for more information
+    # on why we need to include these.
+    ++ lib.optionals
+      (stdenv.cc.isClang || (stdenv.cc.isGNU && lib.versionAtLeast stdenv.cc.cc.version "11.0.0"))
+      [ "-include" "limits" "-include" "thread" ];
+  MALLOC_OVERRIDE = "no";
+  SSLINCDIR = "${lib.getDev opensslStatic}/include";
+  SSLLIBDIR = "${lib.getLib opensslStatic}/lib";
+  MACOSX_SDK = lib.optionalString stdenv.isDarwin "${emptyDirectory}";
+  "LIBC++DIR" = lib.optionalString stdenv.isDarwin "${libcxxUnified}/lib";
 
-  nativeBuildInputs = [ autoPatchelfHook ];
+  jamFlags = [ "-sEXEC=bin.unix" ]
+    ++ lib.optionals stdenv.cc.isClang [ "-sOSCOMP=clang" "-sCLANGVER=${stdenv.cc.cc.version}" ]
+    ++ lib.optionals stdenv.cc.isGNU [ "-sOSCOMP=gcc" "-sGCCVER=${stdenv.cc.cc.version}" ]
+    ++ lib.optionals stdenv.isLinux [ "-sOSVER=26" ]
+    ++ lib.optionals stdenv.isDarwin [ "-sOSVER=1013" ];
+
+  outputs = [ "bin" "dev" "out" ];
```

```suggestion
  outputs = [ "out" "bin" "dev" ];
```
otherwise leftover files end up in bin